Phase 2 tau iso training

fojensen opened the pull request:
The script:
(cmsRun) RecoTauTag/RecoTau/test/rerunMVAIsolationOnMiniAOD_Phase2.py
calculates the bdt score and embeds it in a new tau collection, then runs the analyzer:
RecoTauTag/RecoTau/plugins/rerunMVAIsolationOnMiniAOD_Phase2.cc
to make simple plots.
These two files are roughly analogous to the the Run2 versions:
RecoTauTag/RecoTau/test/rerunMVAIsolationOnMiniAOD.py
/RecoTauTag/RecoTau/test/rerunMVAIsolationOnMiniAOD.cc

As of this PR, the db file:
RecoTauTag/RecoTau/data/RecoTauTag_MVAs_2020Mar25.db 
is loaded locally:
RecoTauTag/Configuration/python/loadRecoTauTagMVAsFromPrepDB_cfi.py

swozniewski commented on RecoTauTag/Configuration/python/loadRecoTauTagMVAsFromPrepDB_cfi.py:
```diff
@@ -297,6 +315,36 @@
+# MVAIso Phase2
+import os
+loadRecoTauTagMVAsFromSQLiteDB_phase2 = loadRecoTauTagMVAsFromPrepDB.clone(
+    connect = 'sqlite_file:'+os.getenv('CMSSW_BASE')+'/src/RecoTauTag/RecoTau/data/RecoTauTag_MVAs_2020Mar25.db',
```
Please switch and load the payloads from central DB.

fojensen replied:
I have switched to the central DB.

swozniewski commented on RecoTauTag/RecoTau/plugins/rerunMVAIsolationOnMiniAOD_Phase2.cc:
```diff
@@ -0,0 +1,226 @@
+#include "FWCore/Framework/interface/EDAnalyzer.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+//#include "DataFormats/PatCandidates/interface/CompositeCandidate.h"
```
please remove if not needed. If the comment is needed please add a comment why.

fojensen replied:
I have removed this line.

swozniewski commented on RecoTauTag/RecoTau/plugins/rerunMVAIsolationOnMiniAOD_Phase2.cc:
```diff
@@ -0,0 +1,226 @@
+
+   bool haveGenJets;
+   edm::EDGetTokenT<std::vector<reco::GenJet>> genJetToken_;   
+   //bool haveGenVisTaus;
```
also here, please remove if not needed

fojensen replied:
I now use the gen visible tau collection made from this producer:
https://github.com/cms-tau-pog/cmssw/blob/CMSSW_11_1_0_X_tau-pog_phase2-mvaiso/PhysicsTools/JetMCAlgos/python/TauGenJets_cfi.py
and therefore have reinstated the genVisTau matching.

mbluj replied:
OK, great, it is what I wanted to suggest. You should however note that it builds genTaus also for tau->e/mu decays, which probably should be excluded here, see e.g. how it is used in PAT/miniAOD: https://github.com/cms-tau-pog/cmssw/blob/CMSSW_11_1_0_X_tau-pog_phase2-mvaiso/PhysicsTools/PatAlgos/python/producersLayer1/tauProducer_cff.py#L14-L15 and [L27-28](https://github.com/cms-tau-pog/cmssw/blob/CMSSW_11_1_0_X_tau-pog_phase2-mvaiso/PhysicsTools/PatAlgos/python/producersLayer1/tauProducer_cff.py#L27-L28).
You can also consider to use patTau.genJet() which receives pointer to matched tauVisJet (made as mentioned above) if found.

fojensen replied:
Please correct me if I am mistaken:
Instead of explicitly producing the visible tau collection myself, I can simply ask if tau->genJet() is non-empty. Importantly, this genJet() only consists of hadronic tau decays.

That is seen in this list of tasks:
https://github.com/cms-tau-pog/cmssw/blob/CMSSW_11_1_0_X_tau-pog_phase2-mvaiso/PhysicsTools/PatAlgos/python/producersLayer1/tauProducer_cff.py#L26-L29

mbluj replied:
Yes, it should work. The only issue is that matching to gen in patTaus is tighter than usually used (dR<0.1 & dPt/Pt<3 vs dR<0.2), see https://github.com/cms-tau-pog/cmssw/blob/CMSSW_11_1_0_X_tau-pog_phase2-mvaiso/PhysicsTools/PatAlgos/python/mcMatchLayer0/tauMatch_cfi.py#L21-L31. But, maybe for testing purposes it is not a real issue.

swozniewski commented on RecoTauTag/RecoTau/plugins/rerunMVAIsolationOnMiniAOD_Phase2.cc:
```diff
@@ -0,0 +1,226 @@
+      haveGenJets = true;
+      genJetToken_ = consumes<std::vector<reco::GenJet>>(iConfig.getParameter<edm::InputTag>("genJetCollection"));
+   }
+   //haveGenVisTaus = false;
```
see above

fojensen replied:
I now use the gen visible tau collection made from this producer:
https://github.com/cms-tau-pog/cmssw/blob/CMSSW_11_1_0_X_tau-pog_phase2-mvaiso/PhysicsTools/JetMCAlgos/python/TauGenJets_cfi.py
and therefore have reinstated the genVisTau matching.

swozniewski commented on RecoTauTag/RecoTau/plugins/rerunMVAIsolationOnMiniAOD_Phase2.cc:
```diff
@@ -0,0 +1,226 @@
+   edm::Handle<std::vector<reco::GenJet>> genJets;
+   if (haveGenJets) iEvent.getByToken(genJetToken_, genJets);
+
+   //edm::Handle<pat::CompositeCandidateCollection> genVisTaus;
```
see above

fojensen replied:
I now use the gen visible tau collection made from this producer:
https://github.com/cms-tau-pog/cmssw/blob/CMSSW_11_1_0_X_tau-pog_phase2-mvaiso/PhysicsTools/JetMCAlgos/python/TauGenJets_cfi.py
and therefore have reinstated the genVisTau matching.

swozniewski commented on RecoTauTag/RecoTau/plugins/rerunMVAIsolationOnMiniAOD_Phase2.cc:
```diff
@@ -0,0 +1,226 @@
+      }
+      if (isPU) continue;
+
+      //bool isSig = false;
```
see above

fojensen replied:
I now use the gen visible tau collection made from this producer:
https://github.com/cms-tau-pog/cmssw/blob/CMSSW_11_1_0_X_tau-pog_phase2-mvaiso/PhysicsTools/JetMCAlgos/python/TauGenJets_cfi.py
and therefore have reinstated the genVisTau matching.

swozniewski commented on RecoTauTag/RecoTau/src/PFRecoTauClusterVariables.cc:
```diff
@@ -192,10 +192,11 @@ namespace reco {
             mvaOpt == kDBnewDMwLTwGJ) &&
            (tauDecayMode == 0 || tauDecayMode == 1 || tauDecayMode == 2 || tauDecayMode == 5 || tauDecayMode == 6 ||
             tauDecayMode == 10 || tauDecayMode == 11))) {
+
```
please drop line insertion

fojensen replied:
Agreed.

swozniewski commented on RecoTauTag/RecoTau/test/rerunMVAIsolationOnMiniAOD_Phase2.py:
```diff
@@ -0,0 +1,136 @@
+)
+setattr(process, "newTauIDsEmbedded", embedID)
+
+## added for mvaIsolation on miniAOD testing
```
please remove if not needed. Same for other commented code below

fojensen replied:
I removed the OutputModule block of code and genVisTauProducer block of code. The genVisTauProducer has been replaced with producer found here:
I now use the gen visible tau collection made from this producer:
https://github.com/cms-tau-pog/cmssw/blob/CMSSW_11_1_0_X_tau-pog_phase2-mvaiso/PhysicsTools/JetMCAlgos/python/TauGenJets_cfi.py

swozniewski commented on RecoTauTag/RecoTau/test/rerunMVAIsolationOnMiniAOD_Phase2.py:
```diff
@@ -0,0 +1,136 @@
+    src = cms.InputTag('slimmedTaus'),
+    tauIDSources = cms.PSet(
+        byIsolationMVAPhase2raw = cms.InputTag('rerunDiscriminationByIsolationMVAPhase2raw'),
+        byVVLooseIsolationMVAPhase2New = cms.InputTag('rerunDiscriminationByIsolationMVAPhase2VVLoose'),
```
Is the 'new' needed in these names? If you want to underline the difference to an older ID, it would be good to add a label that is more explanatory.

mbluj replied:
For consistency with run-2 MVAIso tauIDs I propose to use something like `IsolationMVADBnewDMwLTPhase2` instead of `IsolationMVAPhase2New`

fojensen replied:
"New" is not needed in those in the names, those instances have been removed.

IsolationMVADBnewDMwLTPhase2 will be used as a tag.

swozniewski commented on RecoTauTag/RecoTau/plugins/rerunMVAIsolationOnMiniAOD_Phase2.cc:
```diff
@@ -0,0 +1,226 @@
+#include "FWCore/Framework/interface/EDAnalyzer.h"
```
please move this file to RecoTauTag/RecoTau/test/

fojensen replied:
When rerunMVAIsolationOnMiniAOD_Phase2.cc was in /test it was not compiling, but when it was in /plugins it did. Upon inspection it seems there is a special block in the BuildFile in the test directory for rerunMVAIsolationOnMiniAOD.cc:
https://github.com/cms-tau-pog/cmssw/blob/8c1791c61fce4b13fffae40cc83a2c75dae3469e/RecoTauTag/RecoTau/test/BuildFile.xml#L5-L18
and if add a similiar block to the BuildFile for the rerunMVAIsolationOnMiniAOD_Phase2.cc version it appears to compile properly when in /test.

swozniewski replied:
Alright, please add these as well in this case. Thanks!

mbluj commented on RecoTauTag/RecoTau/interface/PFRecoTauClusterVariables.h:
```diff
@@ -65,7 +65,8 @@ namespace reco {
       kDBoldDMwLTwGJ,
-      kDBnewDMwLTwGJ
+      kDBnewDMwLTwGJ,
+      kPhase2
```
I propose to use more descriptive type similar to run2 ones. e.g. something like `kDBnewDMwLTwGJPhase2` instead of `kPhase2`

fojensen replied:
kDBnewDMwLTwGJPhase2 will be used.

mbluj commented on RecoTauTag/RecoTau/plugins/PATTauDiscriminationByMVAIsolationRun2.cc:
```diff
@@ -110,6 +110,8 @@ namespace reco {
         else if (mvaOpt_string == "DBnewDMwLTwGJ")
           mvaOpt_ = kDBnewDMwLTwGJ;
+        else if (mvaOpt_string == "Phase2")
+          mvaOpt_ = kPhase2;
```
as in other place, use something more descriptive than phase2 here and in other places

fojensen replied:
kDBnewDMwLTwGJPhase2 will be used in its place.

mbluj commented on RecoTauTag/RecoTau/test/rerunMVAIsolationOnMiniAOD_Phase2.py:
```diff
@@ -0,0 +1,136 @@
+process.rerunMVAIsolationOnMiniAOD_Phase2 = cms.EDAnalyzer(
+    'rerunMVAIsolationOnMiniAOD_Phase2',
+    genJetCollection = cms.InputTag("slimmedGenJets"), #comment out to run on data
+    #genVisTauCollection = cms.InputTag("genVisTauProducer:genVisTaus"), #comment out if running on data
```
Will it be useful to add `isData` switch in the analyser and this config to read or not this genVisTau collection? This can be also checked in runtime with `event.isRealData()`, where `event` is `edm::Event& event`.

fojensen replied:
The way the code is setup is that the .cc file explicitly checks if the parameter was specified in the cfg, if it is not specified in the cfg then the analyzer .cc file does not load the collections:
https://github.com/cms-tau-pog/cmssw/blob/8c1791c61fce4b13fffae40cc83a2c75dae3469e/RecoTauTag/RecoTau/plugins/rerunMVAIsolationOnMiniAOD_Phase2.cc#L49-L63

The same concern would not only be for the genVisTau collection, but also for the slimmedGenJets and prunedGenParticles collections:
https://github.com/cms-tau-pog/cmssw/blob/8c1791c61fce4b13fffae40cc83a2c75dae3469e/RecoTauTag/RecoTau/test/rerunMVAIsolationOnMiniAOD_Phase2.py#L123-L125

With that said, it would be even safer to encapsulate the loading of these collections in the .cc file within a isRealData flag. I like this idea -- thank you.

The genVisTau collection is a collection I make with a separate producer which I did not include in this PR, that is why all references to it in this PR are commented out.
In fact, now I remember Anne-Catherine told me about the gen vis tau collection available here:
https://github.com/cms-tau-pog/cmssw/blob/CMSSW_11_1_0_X_tau-pog_phase2-mvaiso/PhysicsTools/JetMCAlgos/python/TauGenJets_cfi.py
Which I have switched to; all references to the prunedGenParticles collection has be removed.

mbluj replied:
OK, it makes sense to me. Please see also my comment above on `TauGenJets`.

fojensen commented:
>  One major item is that the ID should also be added to the official workflow for slimmed taus using an phase2 era-modifier. This can be done similarly to deepTau implementation here
> https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py#L349-L367
> It uses the helper class in https://github.com/cms-sw/cmssw/blob/master/RecoTauTag/RecoTau/python/tools/runTauIdMVA.py but as this might become tricky in combination with an era modifier it would also be fine to directly define and add the ID module in miniAOD_tools.py
> You can find an example for the usage of the modifier in
> https://github.com/cms-sw/cmssw/blob/master/RecoTauTag/Configuration/python/HPSPFTaus_cff.py#L147-L148


FYI: my code runs on Run2 MC as well, it does not need anything unique from Phase2. Therefore it doesn't strictly that it have any sort of Era modifier.

What are the appropriate test scripts I should run for the integration with the official tau workflow? runTheMatrix?

mbluj commented:
> 
> 
> > One major item is that the ID should also be added to the official workflow for slimmed taus using an phase2 era-modifier. This can be done similarly to deepTau implementation here
> > https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py#L349-L367
> > It uses the helper class in https://github.com/cms-sw/cmssw/blob/master/RecoTauTag/RecoTau/python/tools/runTauIdMVA.py but as this might become tricky in combination with an era modifier it would also be fine to directly define and add the ID module in miniAOD_tools.py
> > You can find an example for the usage of the modifier in
> > https://github.com/cms-sw/cmssw/blob/master/RecoTauTag/Configuration/python/HPSPFTaus_cff.py#L147-L148
> 
> FYI: my code runs on Run2 MC as well, it does not need anything unique from Phase2. Therefore it doesn't strictly that it have any sort of Era modifier.

Yes, we understand it, but is there any solid motovation to have it in run2/run3 workflows?

> What are the appropriate test scripts I should run for the integration with the official tau workflow? runTheMatrix?

Yes, in principle you should run `runTheMatrix.py -l limited -i all --ibeos` as explained at: http://cms-sw.github.io/PRWorkflow.html. But, it is the last check to be done, while for now it will be enough that you run just miniAOD production on top of AOD. You can get miniAOD with cmsDriver.py command as:
```
cmsDriver.py step1 --filein "dbs:/VBFHToTauTau_M125_14TeV_powheg_pythia8_correctedGridpack/PhaseIIMTDTDRAutumn18DR-PU200_103X_upgrade2023_realistic_v2-v1/FEVT" --fileout file:HIG-PhaseIIMTDTDRAutumn18MiniAOD-00009.root --mc --eventcontent MINIAODSIM --runUnscheduled --datatier MINIAODSIM --conditions 103X_upgrade2023_realistic_v2 --step PAT --nThreads 8 --geometry Extended2023D35 --era Phase2C4_timing_layer_bar 
```
This particular one is taken from `PhaseIIMTDTDRAutumn18` campaign with CMSSW_10_4_0_mtd5 and requires adaptations to CMSSW_11_1_X (global tag, era, geometry, input file(s), etc.), but I hope it can serve as a good starting point.

fojensen commented:
> Yes, we understand it, but is there any solid motovation to have it in run2/run3 workflows?

There is no reason to have it in the Run2/Run3 workflow.

fojensen commented:
>  I already rebased your development and pushed it to the following branch: https://github.com/cms-tau-pog/cmssw/tree/CMSSW_11_1_0_X_tau-pog_phase2-mvaiso_rebased
> Therefore, please do not update this PR but checkout that branch for further commits and create a new PR to it. I also removed the db file from this branch in order to keep the git history clean from large data files.

Is there a nominal CMSSW release I should be working with? I am getting compilation errors with `CMSSW_11_1_0_pre6`. The first few lines are:
```
/uscms_data/d3/fojensen/integration3/CMSSW_11_1_0_pre6/src/DQMServices/Core/src/ROOTFilePB.pb.h:12:2: error: #error This file was generated by a newer version of protoc which is
 #error This file was generated by a newer version of protoc which is
```

swozniewski commented:
pre7 seems to be available and might work. To be on the safe side, you can also use the latest nightly build of CMSSW_11_1_X (use `scram list` to show available builds, e.g. CMSSW_11_1_X_2020-05-08-2300) and rebase the development branch to official-cmssw/master. This is what I usually do. The drawback for longer developments is that after one or two weeks (I'm not sure about the exact timespan) the build won't be available anymore and cmsenv will fail such that one has to checkout a new build. Finally, in case of merge conflicts with the tau-pog branch, please let me know and I will rebase it as well.

swozniewski commented:
> > Yes, we understand it, but is there any solid motovation to have it in run2/run3 workflows?
> 
> There is no reason to have it in the Run2/Run3 workflow.

Then I would vote for using the modifier.

fojensen commented:
> One major item is that the ID should also be added to the official workflow for slimmed taus using an phase2 era-modifier. This can be done similarly to deepTau implementation here
> https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py#L349-L367
> It uses the helper class in https://github.com/cms-sw/cmssw/blob/master/RecoTauTag/RecoTau/python/tools/runTauIdMVA.py but as this might become tricky in combination with an era modifier it would also be fine to directly define and add the ID module in miniAOD_tools.py
> You can find an example for the usage of the modifier in
> https://github.com/cms-sw/cmssw/blob/master/RecoTauTag/Configuration/python/HPSPFTaus_cff.py#L147-L148

I am sorry, I am maybe a little lost.
Is the idea to modify [runTauIdMVA.py](https://github.com/cms-sw/cmssw/blob/master/RecoTauTag/RecoTau/python/tools/runTauIdMVA.py) to include a line for the Phase2 training. Then, if using a Phase2 modifier, also include the Phase2 training in the "toKeep" box in miniAOD_tools?
https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py#L347

OR, am I too make an identical block similiar to DeepTau, placed immediately below, which is called if a Phase2 modifier exists:
https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py#L339-L357
which takes the slimmedTausDeepIDsv2p1 as an input, adds the training, and outputs a collection called slimmedTausDeepIDsv2p1?

mbluj commented:
> 
> 
> > One major item is that the ID should also be added to the official workflow for slimmed taus using an phase2 era-modifier. This can be done similarly to deepTau implementation here
> > https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py#L349-L367
> > It uses the helper class in https://github.com/cms-sw/cmssw/blob/master/RecoTauTag/RecoTau/python/tools/runTauIdMVA.py but as this might become tricky in combination with an era modifier it would also be fine to directly define and add the ID module in miniAOD_tools.py
> > You can find an example for the usage of the modifier in
> > https://github.com/cms-sw/cmssw/blob/master/RecoTauTag/Configuration/python/HPSPFTaus_cff.py#L147-L148
> 
> I am sorry, I am maybe a little lost.
> Is the idea to modify [runTauIdMVA.py](https://github.com/cms-sw/cmssw/blob/master/RecoTauTag/RecoTau/python/tools/runTauIdMVA.py) to include a line for the Phase2 training. Then, if using a Phase2 modifier, also include the Phase2 training in the "toKeep" box in miniAOD_tools?
> https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py#L347
> 
> OR, am I too make an identical block similiar to DeepTau, placed immediately below, which is called if a Phase2 modifier exists:
> https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py#L339-L357
> which takes the slimmedTausDeepIDsv2p1 as an input, adds the training, and outputs a collection called slimmedTausDeepIDsv2p1?

In my opinion you should:
1. Add the phase2 MVAIso (raw+WPs) to the [runTauIdMVA.py](https://github.com/cms-sw/cmssw/blob/master/RecoTauTag/RecoTau/python/tools/runTauIdMVA.py) tool, so users can easly add it to their workflows on top of miniAOD regardless if the MVAIso is already present in miniAOD samples or not;
2. Add the phase2 MVAIso (raw+WPs) to the official miniAOD workflow for the phase2 era. It can be done with the [runTauIdMVA.py](https://github.com/cms-sw/cmssw/blob/master/RecoTauTag/RecoTau/python/tools/runTauIdMVA.py) tool or without it depending on your convenience.  Adding it without the tool is probably simpler as does not interfere with deepTau which is added for all earas, but either way is fine for us.

swozniewski commented:
True, it would be good to have 1. in any case.
The modifier acts on elements of the cmssw config and the tauIDEmbedder acts on the inserted process right away, so I'm not sure whether it is doable or possible to use both together. If you have an idea, you're welcome to use it of course. Otherwise it's probably easier to add a seperate block without using the tauIDEmbedder.

mbluj commented:
> 
> 
> True, it would be good to have 1. in any case.
> The modifier acts on elements of the cmssw config and the tauIDEmbedder acts on the inserted process right away, so I'm not sure whether it is doable or possible to use both together. If you have an idea, you're welcome to use it of course. Otherwise it's probably easier to add a seperate block without using the tauIDEmbedder.

It is doable, but a bit tricky, see what is done for deepTau in 10_6_X:
1. for miniAOD: https://github.com/cms-sw/cmssw/blob/CMSSW_10_6_X/PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py#L366-L383
2. for nanoAOD: https://github.com/cms-sw/cmssw/blob/CMSSW_10_6_X/PhysicsTools/NanoAOD/python/nano_cff.py#L148-L156 and then in a common customization function in L303. (You can also check this for 10_2_X with a bit different implementation)

fojensen commented:
> > What are the appropriate test scripts I should run for the integration with the official tau workflow? runTheMatrix?
> 
> Yes, in principle you should run `runTheMatrix.py -l limited -i all --ibeos` as explained at: http://cms-sw.github.io/PRWorkflow.html. But, it is the last check to be done, while for now it will be enough that you run just miniAOD production on top of AOD. You can get miniAOD with cmsDriver.py command as:
> 
> ```
> cmsDriver.py step1 --filein "dbs:/VBFHToTauTau_M125_14TeV_powheg_pythia8_correctedGridpack/PhaseIIMTDTDRAutumn18DR-PU200_103X_upgrade2023_realistic_v2-v1/FEVT" --fileout file:HIG-PhaseIIMTDTDRAutumn18MiniAOD-00009.root --mc --eventcontent MINIAODSIM --runUnscheduled --datatier MINIAODSIM --conditions 103X_upgrade2023_realistic_v2 --step PAT --nThreads 8 --geometry Extended2023D35 --era Phase2C4_timing_layer_bar 
> ```
> 
> This particular one is taken from `PhaseIIMTDTDRAutumn18` campaign with CMSSW_10_4_0_mtd5 and requires adaptations to CMSSW_11_1_X (global tag, era, geometry, input file(s), etc.), but I hope it can serve as a good starting point.

I am having difficulty with the cmsDriver command. I thought a good starting point would be to find a good Phase 2 MINIAODSIM and then use the setup command obtained through MCM to produce the MINIAODSIM from the GEN-SIM-DIGI-RAW.

`/VBFHToTauTau_M125_TuneCUETP8M1_14TeV_powheg_pythia8/Phase2HLTTDRWinter20RECOMiniAOD-PU140_110X_mcRun4_realistic_v3-v1/MINIAODSIM `

`cmsDriver.py step1 --filein "dbs:/VBFHToTauTau_M125_TuneCUETP8M1_14TeV_powheg_pythia8/Phase2HLTTDRWinter20DIGI-PU140_110X_mcRun4_realistic_v3-v1/GEN-SIM-DIGI-RAW" --fileout file:TSG-Phase2HLTTDRWinter20RECOMiniAOD-00073.root --mc --eventcontent MINIAODSIM --runUnscheduled --customise SLHCUpgradeSimulations/Configuration/aging.customise_aging_1000,Configuration/DataProcessing/Utils.addMonitoring --datatier MINIAODSIM --conditions 110X_mcRun4_realistic_v3 --step RAW2DIGI,L1Reco,RECO,RECOSIM,PAT --nThreads 8 --geometry Extended2026D49 --era Phase2C9 --python_filename TSG-Phase2HLTTDRWinter20RECOMiniAOD-00073_1_cfg.py --no_exec -n 52`

This works in `CMSSW_11_0_0_patch1` but fails inside `CMSSW_11_1_0_pre7` with an error:
...

```
[18] Calling method for module SiPixelRecHitConverter/'siPixelRecHits'
Exception Message:
SiPixelGenError::qbin can't find needed GenError ID = 0
```

mbluj commented:
Hi Frank,
I deduced from matrix tests the following cmsDriver command to test miniAOD on top of AOD/RECO:
```
cmsDriver miniAOD2026 --filein root://eoscms.cern.ch//eos/cms/store/relval/CMSSW_11_1_0_pre6/RelValTTbar_14TeV/GEN-SIM-RECO/PU25ns_110X_mcRun4_realistic_v3_2026D49PU200-v1/20000/FF9B5B74-B6D3-0746-92FB-C2F57C4DF6DC.root --fileout file:MiniAOD2026.root --mc --eventcontent MINIAODSIM --runUnscheduled --datatier MINIAODSIM --conditions auto:phase2_realistic_T14 --step PAT --nThreads 4 --geometry Extended2026D49 --scenario pp --era Phase2C8 -n -1 --no_exec
```
Used file consists of 100 events which number should be fine for functional tests, but if you need more there are several files with 9k events in total. You can also try pre7 files from analogous location (pre6->pre7).
Good luck!

mbluj commented:
@fojensen: Hi Frank - how are you? Do you progress? 
We are being pinged by Tau POG conveners (who are in turn pinged by higher-level coordinators) to have finally phase-2 specialized tau-IDs available for users. Therefore, we would like to hear from you about realistic plan to successfully finalize this development (which was initially expected in December 2019).

fojensen commented:
@mbluj 
I have added the appropriate lines to runTauIdMVA.py to allow a user to add the training on the fly if they wish.
The last thing I need to do is figure out how to incorporate the training into the official workflow. You mentioned to add it without runTauIdMVA/TauIDEmbedder... Do you have an example of adding a training without these tools?

> On May 26, 2020, at 9:53 AM, mbluj <notifications@github.com> wrote:
> 
> @fojensen: Hi Frank - how are you? Do you progress?
> We are being pinged by Tau POG conveners (who are in turn pinged by higher-level coordinators) to have finally phase-2 specialized tau-IDs available for users. Therefore, we would like to hear from you about realistic plan to successfully finalize this development (which was initially expected in December 2019).

swozniewski commented:
Hi Frank, I think it's basically what you already do in rerunMVAIsolationOnMiniAOD_Phase2.py. I.e. first add you producer to the process and the 'task' (as also done for other producers in minoAOD_tools.py). Then you need to register the products in the tauIDEmbedder such that it is added to the resulting PATTaus. Here, the tauIDEmbedder is already added by the helper class as  process.slimmedTaus. I think it should work to append your IDs to its tauIDSources attribute. This would keep the other IDs, which we want at the moment, right? @mbluj may correct me

mbluj commented:
> @mbluj I have added the appropriate lines to runTauIdMVA.py to allow a user to add the training on the fly if they wish. 

This is great! Is it already committed?

> The last thing I need to do is figure out how to incorporate the training into the official workflow. You mentioned to add it without runTauIdMVA/TauIDEmbedder... Do you have an example of adding a training without these tools?

You should always use TauIDEmbedder module, but it is not mandatory to use the runTauIdMVA tool. Instead, you can add required modules running MVAIso by hand to miniAOD configuration and then to an tauIDEmbedder module which is already present in miniAOD. See here
https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py:
* in [L346](https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py#L346) tauIDEmbedder module used for embed deepTau is renamed to `slimmedTaus`
* starting from L351 you define set of modules to include (actually two modules one defining raw discriminator and other its WPs) and then a task containing them. The new task should be added to the main task (or to the deepTauID task) for phase2 era using appropriate modifier. You should note that source tau for MVAIso should be called `_noUpdatedTauName`
* then you can define a PSet with the new modules to be added to the tauIDEmbedded, more or less like this:
```
_phase2tauMvaIsoIdPSet = cms.PSet(
process.slimmedTaus.tauIDSources,
 your PSets with Raw and WPs of MVAIso
)
```
to replace to `process.slimmedTaus.tauIDSources` using era modifier.

Is it more or less clear?

Alternatively, you can try play with the runTauIdMVA tool, but I think it is more tricky to use it together with era modifiers than direct way briefly described above.

mbluj commented:
> 
> 
> Hi Frank, I think it's basically what you already do in rerunMVAIsolationOnMiniAOD_Phase2.py. I.e. first add you producer to the process and the 'task' (as also done for other producers in minoAOD_tools.py). Then you need to register the products in the tauIDEmbedder such that it is added to the resulting PATTaus. Here, the tauIDEmbedder is already added by the helper class as process.slimmedTaus. I think it should work to append your IDs to its tauIDSources attribute. This would keep the other IDs, which we want at the moment, right? @mbluj may correct me

Yes, correct. It is what I tried to describe also in my answer.

fojensen commented:
> This is great! Is it already committed?

I have things on my local branch:
https://github.com/fojensen/cmssw/tree/CMSSW_11_1_0_X_tau-pog_phase2-mvaiso_rebased
where I have all the work done up to adding the training to Phase2 CMSSW.

fojensen commented:
I have created a new PR:
https://github.com/cms-tau-pog/cmssw/pull/139

Where all discussions for this PR have been taken in to account.

mbluj commented:
@swozniewski, can now we close this PR?

swozniewski commented:
yes, goog point
